### PR TITLE
Save json report to file

### PR DESCRIPTION
**Pull request by NickTroy**
Hello!

I need to post-process json data generated in report which is not possible because now json data is printed to console. That's why I added a feature that creates a file 'report.json' in Config.root directory and saves json data there.

Tell me whether this feature needs testing.

With respect, Mykyta Troianov.

### CODE REVIEW

File: lib/rubycritic/generators/json/simple.rb [hunk 1]
```diff
@@ -24,6 +25,18 @@ def data
...
+        def file_directory
+          @file_directory ||= Pathname.new(Config.root)
+        end
+
+        def file_name
+          'report.json'
+        end
+
+        def file_pathname
+          File.join(file_directory, file_name)
```
Comment (troessner) [inline lib/rubycritic/generators/json/simple.rb]: Why use File here and not Pathname? Also I believe you can merge `file_name` and `file_pathname`, this can be one method.

File: lib/rubycritic/generators/json/simple.rb [hunk 2]
```diff
@@ -24,6 +25,18 @@ def data
             score: @analysed_modules.score
           }
         end
+
+        def file_directory
+          @file_directory ||= Pathname.new(Config.root)
+        end
+
+        def file_name
+          'report.json'
```
Comment (troessner) [inline lib/rubycritic/generators/json/simple.rb]: Let's make this a constant.

File: lib/rubycritic/generators/json/simple.rb [hunk 3]
```diff
@@ -24,6 +25,18 @@ def data
             score: @analysed_modules.score
           }
         end
+
+        def file_directory
+          @file_directory ||= Pathname.new(Config.root)
```
Comment (troessner) [inline lib/rubycritic/generators/json/simple.rb]: I think this should be configurable and not kind of hard-coded.

**Comment by troessner** [comment]
I like the proposal but I believe this needs some polishing before this can go live.
I see 2 big issues:

1.) No specs
2.) The path directory should be configurable because otherwise this will pollute the project directory which will alienate users

Apart from that please see my other comments.

Let's get this going!

File: lib/rubycritic/generators/json/simple.rb (hunk 1 above)
Comment (NickTroy) [inline lib/rubycritic/generators/json/simple.rb]: Merged #file_name and #file_pathname to #file_pathname. I used File.join because it looks simpler to me, can substitute it with  Pathname.new(file_directory, FILE_NAME), should I?

File: lib/rubycritic/generators/json/simple.rb (hunk 2 above)
Comment (NickTroy) [inline lib/rubycritic/generators/json/simple.rb]: Made it a constant.

File: lib/rubycritic/generators/json/simple.rb (hunk 3 above)
Comment (NickTroy) [inline lib/rubycritic/generators/json/simple.rb]: At the moment user can configure saving directory by using -p "path/to/directory/", this feature is still working.

**Comment by NickTroy** [comment]
Completely agree with you about polishing. But will need your advise on:
1) What exactly should be tested in this feature? As I see it now this should be covered: creating file, checking correctness of directory, checking data inside file.
Did I understand you correct about configurable file directory?

File: lib/rubycritic/generators/json/simple.rb (hunk 3 above)
Comment (troessner) [inline lib/rubycritic/generators/json/simple.rb]: >> At the moment user can configure saving directory by using -p "path/to/directory/", this feature is still working.

Good point, then please just ignore my comment :)

**Comment by troessner** [comment]
>> As I see it now this should be covered: creating file, checking correctness of directory, checking data inside file.

We don't have to be to pedantic about it but I think we should at least check that a non-empty file was generated.

>> Did I understand you correct about configurable file directory?

Yeah, see the comment above, you can ignore this ;)

**Comment by NickTroy** [comment]
> We don't have to be to pedantic about it but I think we should at least check that a non-empty file was generated.
I will add a test that checks for non-empty file report.json in concrete folder.

**Comment by NickTroy** [comment]
Added a test to check whether file exists and whether it's not empty. Advise on adding any other tests, if they are needed.

File: test/lib/rubycritic/generators/json_report_test.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+require 'test_helper'
+require 'rubycritic/core/analysed_modules_collection'
+require 'rubycritic/generators/json_report'
+require 'json'
+
+describe RubyCritic::Generator::JsonReport do
+  describe '#generate_report' do
+    before do
+      FileUtils.rm('test/samples/report.json') if File.file?('test/samples/report.json')
```
Comment (troessner) [inline test/lib/rubycritic/generators/json_report_test.rb]: Judging from experience I'd say that this will turn into a flickering test one day since you're operating on the real filesystem. How about using [fakefs](https://github.com/fakefs/fakefs) to defuse this bomb right from the start?

Comment (troessner) [inline test/lib/rubycritic/generators/json_report_test.rb]: This will also allow us to drop the line above

File: test/lib/rubycritic/generators/json_report_test.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      create_analysed_modules_collection
+      generate_report
```
Comment (troessner) [inline test/lib/rubycritic/generators/json_report_test.rb]: I find it confusing that the examples below just contain "assert". I think we should move the report generation into the examples. Using a fake file system like suggested above will also make this easier since you don't have to clean up afterwards. Meaning that you can just move "generate_report" to each examples with duplicating the set up everywhere.

File: lib/rubycritic/generators/json_report.rb [hunk 6]
```diff
@@ -8,7 +8,10 @@ def initialize(analysed_modules)
       end
 
       def generate_report
-        print generator.render
+        FileUtils.mkdir_p(generator.file_directory)
```
Comment (troessner) [inline lib/rubycritic/generators/json_report.rb]: We can and should use [Pathname](https://ruby-doc.org/stdlib-2.1.0/libdoc/pathname/rdoc/Pathname.html#method-i-mkdir) here.

File: lib/rubycritic/generators/json_report.rb [hunk 7]
```diff
@@ -8,7 +8,10 @@ def initialize(analysed_modules)
...
+        File.open(generator.file_pathname, 'w+') do |file|
```
Comment (troessner) [inline lib/rubycritic/generators/json_report.rb]: And here.

File: lib/rubycritic/generators/json/simple.rb [hunk 8]
```diff
@@ -24,6 +27,14 @@ def data
...
+        end
+
+        def file_pathname
+          File.join(file_directory, FILE_NAME)
```
Comment (troessner) [inline lib/rubycritic/generators/json/simple.rb]: Let's re-write this to

```Ruby
Pathname.new(file_directory).join FILE_NAME
```

File: lib/rubycritic/generators/json_report.rb (hunk 6 above)
Comment (NickTroy) [inline lib/rubycritic/generators/json_report.rb]: This would look more complicated, because #mkdir throws error if folder already exists.
`Pathname.new(generator.file_directory).mkdir` unless File.directory? generator.file_directory

File: lib/rubycritic/generators/json_report.rb (hunk 7 above)
Comment (NickTroy) [inline lib/rubycritic/generators/json_report.rb]: Same as above.

File: test/lib/rubycritic/generators/json_report_test.rb (hunk 4 above)
Comment (NickTroy) [inline test/lib/rubycritic/generators/json_report_test.rb]: Reworked tests to use FakeFS.

File: test/lib/rubycritic/generators/json_report_test.rb (hunk 5 above)
Comment (NickTroy) [inline test/lib/rubycritic/generators/json_report_test.rb]: In recent change I added report is generated before each test using FakeFS.

**Comment by NickTroy** [comment]
Tell me, if you think that we should use Pathname anyway, will rework it asap.

File: lib/rubycritic/generators/json_report.rb (hunk 6 above)
Comment (troessner) [inline lib/rubycritic/generators/json_report.rb]: Alright, fair enough.

**Comment by troessner** [comment]
Looking good now, merged, thanks a lot!

**Comment by NickTroy** [comment]
Was a great experience for me! Thank you as well!

**Comment by NickTroy** [comment]
One final question, when this feature will be published to rubygems?

**Comment by troessner** [comment]
>> One final question, when this feature will be published to rubygems?

2 minutes ago:

>> Tagged v2.9.3.
Pushed git commits and tags.
Pushed rubycritic 2.9.3 to rubygems.org

😆

**Comment by NickTroy** [comment]
Thanks again!)
